pkg/actors/actors.go

```diff
@@ -15,1431 +15,554 @@ package actors
-		return err
-	}
-
-	return a.timers.CreateTimer(ctx, reminder)
+	a.registerDoneLock.Lock()
+	a.registerDoneCh = make(chan struct{})
+	a.registerDoneLock.Unlock()
 }
 
-func (a *actorsRuntime) DeleteReminder(ctx context.Context, req *DeleteReminderRequest) error {
-	if !a.actorsConfig.Config.HostedActorTypes.IsActorTypeHosted(req.ActorType) {
-		return ErrReminderOpActorNotHosted
+func (a *actors) WaitForRegisteredHosts(ctx context.Context) error {
+	if err := a.disabled.Load(); err != nil {
+		return *err
 	}
 
-	return a.actorsReminders.DeleteReminder(ctx, *req)
-}
+	a.registerDoneLock.RLock()
+	rch := a.registerDoneCh
+	a.registerDoneLock.RUnlock()
 
-func (a *actorsRuntime) GetReminder(ctx context.Context, req *GetReminderRequest) (*internal.Reminder, error) {
-	if !a.actorsConfig.Config.HostedActorTypes.IsActorTypeHosted(req.ActorType) {
-		return nil, ErrReminderOpActorNotHosted
+	select {
+	case <-rch:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
 	}
-
-	return a.actorsReminders.GetReminder(ctx, req)
-}
-
-func (a *actorsRuntime) DeleteTimer(ctx context.Context, req *DeleteTimerRequest) error {
-	return a.timers.DeleteTimer(ctx, req.Key())
 }
 
-func (a *actorsRuntime) RegisterInternalActor(ctx context.Context, actorType string, factory InternalActorFactory, actorIdleTimeout time.Duration) error {
-	if !a.haveCompatibleStorage() {
-		return fmt.Errorf("unable to register internal actor type '%s': %w", actorType, ErrIncompatibleStateStore)
+func (a *actors) handleIdleActor(target targets.Idlable) {
+	if err := a.placement.Lock(context.Background()); err != nil {
+		return
 	}
+	defer a.placement.Unlock()
 
-	// Call GetOrSet which returns "existing=true" if the actor type was already registered
-	_, existing := a.internalActorTypes.GetOrSet(actorType, factory)
-	if existing {
-		return fmt.Errorf("actor type '%s' already registered", actorType)
-	}
-
-	log.Debugf("Registered internal actor type '%s'", actorType)
+	log.Debugf("Actor %s is idle, deactivating", target.Key())
 
-	a.actorsConfig.Config.HostedActorTypes.AddActorType(actorType, actorIdleTimeout)
-
-	if a.placementEnabled {
-		err := a.placement.AddHostedActorType(actorType, actorIdleTimeout)
-		if err != nil {
-			return fmt.Errorf("error updating hosted actor types: %w", err)
-		}
+	if err := a.table.Halt(context.Background(), target); err != nil {
+		log.Errorf("Failed to halt actor %s: %s", target.Key(), err)
+		return
 	}
-	return nil
 }
 
-func (a *actorsRuntime) GetRuntimeStatus(ctx context.Context) *runtimev1pb.ActorRuntime {
-	// Do not populate RuntimeStatus, which will be populated by the runtime
-	res := &runtimev1pb.ActorRuntime{
-		ActiveActors: a.getActiveActorsCount(ctx),
+func (a *actors) RuntimeStatus() *runtimev1pb.ActorRuntime {
+	select {
+	case <-a.registerDoneCh:
+		if a.disabled.Load() != nil {
+			return &runtimev1pb.ActorRuntime{
+				RuntimeStatus: runtimev1pb.ActorRuntime_DISABLED,
+				Placement:     "placement: disconnected",
```

Comment:
Should this Placement value be different from the one below? As of now they are identical. Maybe we should put the error value in somewhere?

```suggestion
				Placement:     fmt.Sprintf("placement: errored due to %w", a.disabled.Load())
```

Reply:
We can't change the string because I believe it would be changing the API.